File: php-oop/patterns/factory1.php
<?php

// Produkt
interface Button {
    public function render(): string;
}

// Konkretne produkty
class WebButton implements Button {
    public function render(): string {
        return "<button>Web Button</button>";
    }
}

class DesktopButton implements Button {
    public function render(): string {
        return "[Desktop Button]";
    }
}

// Fabryka abstrakcyjna
interface ButtonFactory {
    public function createButton(): Button;
}

// Konkretne fabryki
class WebButtonFactory implements ButtonFactory {
    public function createButton(): Button {
        return new WebButton();
    }
}

class DesktopButtonFactory implements ButtonFactory {
    public function createButton(): Button {
        return new DesktopButton();
    }
}

// Klient
function renderButton(ButtonFactory $factory) {
    $button = $factory->createButton();
    echo $button->render() . "\n";
}

// Testowanie
$webFactory = new WebButtonFactory();
$desktopFactory = new DesktopButtonFactory();

echo "Web Button:\n";
renderButton($webFactory);

echo "\nDesktop Button:\n";
renderButton($desktopFactory);

File: python/102/patterns/factory1.py
from abc import ABC, abstractmethod

# Produkt
class Button(ABC):
    @abstractmethod
    def render(self):
        pass


# Konkretne produkty
class WebButton(Button):
    def render(self):
        return "<button>Web Button</button>"


class DesktopButton(Button):
    def render(self):
        return "[Desktop Button]"


# Fabryka abstrakcyjna
class ButtonFactory(ABC):
    @abstractmethod
    def create_button(self):
        pass


# Konkretne fabryki
class WebButtonFactory(ButtonFactory):
    def create_button(self):
        return WebButton()


class DesktopButtonFactory(ButtonFactory):
    def create_button(self):
        return DesktopButton()


# Klient
def render_button(factory: ButtonFactory):
    button = factory.create_button()
    print(button.render())


# Testowanie
web_factory = WebButtonFactory()
desktop_factory = DesktopButtonFactory()

print("Web Button:")
render_button(web_factory)
